### PR TITLE
restructure

### PR DESCRIPTION
**Pull request by Dekken**
For https://github.com/X-DataInitiative/tick/issues/124

- Separate sources and includes
- Single include directory
- All include paths complete and relative from ./include
- Minimise setup.py - only use search directories
- Move array and array_test up one level from under "base"
- CMake files updated to pass travis build
- all sh files in root moved to ./sh - updated accordingly

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [submodule "tick/third_party/cereal"]
```
Comment (Mbompr) [inline .gitmodules]: not sure but shouldn't this be lib/third_party/cereal ?

File: lib/include/tick/array/carray_python.h [hunk 2]
```diff
@@ -18,7 +18,7 @@
 #include "ssparsearray2d.h"
 #include "sbasearray.h"
 #include "sbasearray2d.h"
-#include "varray.h"
+#include "tick/array/varray.h"
```
Comment (Mbompr) [inline lib/include/tick/array/carray_python.h]: When do we need full path and when can we use file name directly?

File: clean_build_test.sh [hunk 3]
```diff
@@ -1,6 +0,0 @@
-#!/usr/bin/env bash
-
-echo "CLEAN"
-./full_clean.sh
-
```
Comment (Mbompr) [inline clean_build_test.sh]: I think it was cool to have this file even if its content only executes python commands. It is useful when people have problem with their local installation.

File: lib/src/cpp/array/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,23 @@
+add_library(tick_array EXCLUDE_FROM_ALL
```
Comment (Mbompr) [inline lib/src/cpp/array/CMakeLists.txt]: Can we define at the top of this file:
set(ARRAY_INCLUDE_DIR, "${PROJECT_SOURCE_DIR}/lib/include/tick/array/")
?

File: lib/src/swig/CMakeLists.txt [hunk 5]
```diff
@@ -21,8 +21,8 @@ endif()
```
Comment (Mbompr) [inline lib/src/swig/CMakeLists.txt]: Why is the main CMakeLists in the swig folder?

File: sh/swig.sh [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/usr/bin/env bash-Iinclude
```
Comment (Mbompr) [inline sh/swig.sh]: typo?

File: tick/base/__init__.py [hunk 7]
```diff
@@ -18,7 +18,7 @@ def _set_mpl_backend():
 
 _set_mpl_backend()
 
-from .array import *
+from tick.array import *
```
Comment (Mbompr) [inline tick/base/__init__.py]: Why not but was this necessary?

File: tick/plot/tests/plot_hawkes_test.py [hunk 8]
```diff
@@ -1,67 +0,0 @@
-# License: BSD 3 clause
```
Comment (Mbompr) [inline tick/plot/tests/plot_hawkes_test.py]: You have removed test folders for plot

File: .gitmodules (hunk 1 above)
Comment (Dekken) [reply .gitmodules]: probably, not sure it matters but I'll change it for consistency

File: clean_build_test.sh (hunk 3 above)
Comment (Dekken) [reply clean_build_test.sh]: so what, remove it?

File: lib/src/cpp/array/CMakeLists.txt (hunk 4 above)
Comment (Dekken) [reply lib/src/cpp/array/CMakeLists.txt]: we can - you're the CMake guy! ^_^

File: lib/src/swig/CMakeLists.txt (hunk 5 above)
Comment (Dekken) [reply lib/src/swig/CMakeLists.txt]: Copy paste - will remove

File: sh/swig.sh (hunk 6 above)
Comment (Dekken) [reply sh/swig.sh]: :|


yes

File: tick/base/__init__.py (hunk 7 above)
Comment (Dekken) [reply tick/base/__init__.py]: array is no longer a subdirectory of base so I believe so

File: tick/plot/tests/plot_hawkes_test.py (hunk 8 above)
Comment (Dekken) [reply tick/plot/tests/plot_hawkes_test.py]: whoops

**Comment by Mbompr** [comment]
Warning each time we run setup.py
UserWarning: Included file t2exp.inl in folder lib/src/cpp/base/math has an unknown extension ".inl"
Looks like this file extension is not handled

**Comment by Dekken** [comment]
https://github.com/X-DataInitiative/tick/blob/master/setup.py#L256

```
                if ext == '.inl':
                    pass
```

File: lib/src/cpp/array/CMakeLists.txt (hunk 4 above)
Comment (Dekken) [reply lib/src/cpp/array/CMakeLists.txt]: Done
https://github.com/Dekken/tick/blob/re-org/lib/src/cpp/array/CMakeLists.txt

File: lib/include/tick/array/alloc.h [hunk 9]
```diff
@@ -5,7 +5,7 @@
 
 /** @file */
 
-#include "debug.h"
+#include "tick/base/debug.h"
```
Comment (Mbompr) [inline lib/include/tick/array/alloc.h]: is full path necessary or a good practice or a find/replace mistake?

File: lib/src/cpp/array/CMakeLists.txt (hunk 4 above)
Comment (Mbompr) [reply lib/src/cpp/array/CMakeLists.txt]: I've done it for other CMakeLists files as well

File: lib/src/cpp/base/math/t2exp.inl [hunk 10]
```diff
@@ -0,0 +1,2078 @@
+/*
```
Comment (Mbompr) [inline lib/src/cpp/base/math/t2exp.inl]: This file exists at two locations:
* lib/src/cpp/base/math/t2exp.inl
* lib/include/tick/base/math/t2exp.inl
One should probably be removed

File: tick/inference/inference.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
+# License: BSD 3 clause
```
Comment (Mbompr) [inline tick/inference/inference.py]: This file is a duplication of tick/inference/tests/inference.py and should be deleted

File: tick/optim/model/generalized_linear_model.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
+# License: BSD 3 clause
```
Comment (Mbompr) [inline tick/optim/model/generalized_linear_model.py]: This file is a duplication of  tick/optim/model/tests/generalized_linear_model.py and should be deleted

File: tick/optim/model/hawkes_utils.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
+# License: BSD 3 clause
```
Comment (Mbompr) [inline tick/optim/model/hawkes_utils.py]: This file is a duplication of tick/optim/model/hawkes_utils.py and should be deleted

File: lib/include/tick/array/alloc.h (hunk 9 above)
Comment (Mbompr) [reply lib/include/tick/array/alloc.h]: Ok now full path must be used everywhere, the only exception is when two .h files are in the same folder.

**Comment by Dekken** [comment]
Auto closed on commit?

**Pull request merged**